src/vs/platform/mcp/node/mcpManagementService.ts

```diff
@@ -21,10 +25,117 @@ export class McpUserResourceManagementService extends CommonMcpUserResourceManag
+			if (!local) {
+				throw new Error(`Failed to install MCP server: ${server.name}`);
+			}
+			return local;
+		} catch (e) {
+			this._onDidInstallMcpServers.fire([{ name: server.name, source: server, error: e, mcpResource: this.mcpResource }]);
+			throw e;
+		}
+	}
+
+	private async getNodePackageMcpServerConfiguration(server: IGalleryMcpServer, manifest: IMcpServerManifest): Promise<Omit<IInstallableMcpServer, 'name'>> {
+		if (this.configurationService.getValue('chat.mcp.enableNpmInstall') !== true) {
+			return this.getMcpServerConfigurationFromManifest(manifest, PackageType.NODE);
+		}
+
+		const nodePackage = manifest.packages?.find(p => p.registry_name === PackageType.NODE);
+		if (!nodePackage) {
+			throw new Error('No npm package found in manifest');
+		}
+
+		const location = this.getLocation(server.name, server.version);
+		if (!(await this.fileService.exists(location))) {
+			await this.fileService.createFolder(location);
+		}
+
+		const packageVersion = !nodePackage.version || nodePackage.version === 'latest' ? await this.npmPackageService.getLatestPackageVersion(nodePackage.name) : nodePackage.version;
+		await this.npmPackageService.installPackage(nodePackage.name, packageVersion, location.with({ scheme: Schemas.file }).fsPath);
+
+		const entryPoint = await this.findPackageEntryPoint(nodePackage.name, location);
+		const result = this.getMcpServerConfigurationFromManifest(manifest, PackageType.NODE);
+		if (result.config.type === McpServerType.LOCAL && result.config.args) {
+			const newArgs = [...result.config.args];
+			newArgs[0] = entryPoint.with({ scheme: Schemas.file }).fsPath;
+			return {
+				config: {
+					...result.config,
+					command: 'node',
+					args: newArgs,
+					version: server.version
+				},
+				inputs: result.inputs
+			};
+		}
+		return result;
+	}
+
+	private async findPackageEntryPoint(packageName: string, packageLocation: URI): Promise<URI> {
+		const nodeModulesPath = this.uriIdentityService.extUri.joinPath(packageLocation, 'node_modules', packageName);
+		const packageJsonPath = this.uriIdentityService.extUri.joinPath(nodeModulesPath, 'package.json');
+
+		try {
+			const packageJsonContent = await this.fileService.readFile(packageJsonPath);
+			const packageJson = JSON.parse(packageJsonContent.value.toString());
+
+			if (packageJson.bin) {
+				if (typeof packageJson.bin === 'string') {
+					return this.uriIdentityService.extUri.joinPath(nodeModulesPath, packageJson.bin);
+				} else if (typeof packageJson.bin === 'object') {
+					const binName = packageJson.bin[packageName] || Object.values(packageJson.bin)[0];
+					if (binName) {
+						return this.uriIdentityService.extUri.joinPath(nodeModulesPath, binName as string);
+					}
+				}
+			}
+
+			if (packageJson.main) {
+				return this.uriIdentityService.extUri.joinPath(nodeModulesPath, packageJson.main);
+			}
+
+			return this.uriIdentityService.extUri.joinPath(nodeModulesPath, 'index.js');
```

Comment:
I suggest checking `packageJson.exports?.['.']` too for packages using export maps